File: scripts/seed.ts
import { PrismaClient } from "@prisma/client";
import bcrypt from "bcryptjs";

const prisma = new PrismaClient();

// Bricks name -> ZenHR employment_number, for the employees whose name in
// Bricks (sales visits app) doesn't match their ZenHR name. Used in phase 2
// to link visits back to the right employee. Safe to seed now.
const NAME_OVERRIDES: { employmentNumber: string; bricksDisplayName: string }[] = [
  { employmentNumber: "115", bricksDisplayName: "Kareem Elsobky" }, // ZenHR: Karim Ahmed
  { employmentNumber: "104", bricksDisplayName: "Kareem Sherif" },
  { employmentNumber: "127", bricksDisplayName: "Kareim Helmy" },
  { employmentNumber: "135", bricksDisplayName: "Amgad" },
  { employmentNumber: "119", bricksDisplayName: "Mohamed Abdelrazek" },
  { employmentNumber: "136", bricksDisplayName: "Amr Mostafa" },
  { employmentNumber: "205", bricksDisplayName: "Safaa Sabry" },
  { employmentNumber: "212", bricksDisplayName: "Abdullah Ibrahim Habashi" },
];

// Job role -> dashboard category, seeded from the real job roles seen in
// the business's ZenHR attendance export (Management & Warehouse block vs.
// Sales, collectors & Logistics block). Add a row here whenever a new job
// role shows up as OTHER on the dashboard.
const JOB_ROLE_CATEGORIES: { jobRole: string; category: "MANAGEMENT" | "SALES_COLLECTOR" | "DELIVERY_AGENT" }[] = [
  // Management & Warehouse - measured by hours worked
  { jobRole: "Managing Director", category: "MANAGEMENT" },
  { jobRole: "Supply  chain Manager", category: "MANAGEMENT" },
  { jobRole: "Finance Manager", category: "MANAGEMENT" },
  { jobRole: "Warehouse Supervisor", category: "MANAGEMENT" },
  { jobRole: "Logistics Supervisor", category: "MANAGEMENT" },
  { jobRole: "Accountant", category: "MANAGEMENT" },
  { jobRole: "Purchasing Manager", category: "MANAGEMENT" },
  { jobRole: "Commercial Director", category: "MANAGEMENT" },
  { jobRole: "HR Executive", category: "MANAGEMENT" },
  { jobRole: "office boy/girl", category: "MANAGEMENT" },
  // Sales & Collectors - measured by Bricks visit count
  { jobRole: "Key Account Manager", category: "SALES_COLLECTOR" },
  { jobRole: "Collector", category: "SALES_COLLECTOR" },
  { jobRole: "Senior Account Manager", category: "SALES_COLLECTOR" },
  { jobRole: "Account Manager", category: "SALES_COLLECTOR" },
  { jobRole: "Head of sales", category: "SALES_COLLECTOR" },
  // Delivery Agents - tracked in ZenHR, own pending section (attendance only)
  { jobRole: "Delivery agent", category: "DELIVERY_AGENT" },
];

async function main() {
  for (const override of NAME_OVERRIDES) {
    await prisma.employeeNameOverride.upsert({
      where: { employmentNumber: override.employmentNumber },
      create: override,
      update: override,
    });
  }
  console.log(`Seeded ${NAME_OVERRIDES.length} employee name overrides.`);

  for (const rule of JOB_ROLE_CATEGORIES) {
    await prisma.jobRoleCategoryRule.upsert({
      where: { jobRole: rule.jobRole },
      create: rule,
      update: rule,
    });
  }
  console.log(`Seeded ${JOB_ROLE_CATEGORIES.length} job role category rules.`);

  // Single-admin shortcut, handy for local dev.
  const adminEmail = process.env.SEED_ADMIN_EMAIL;
  const adminPassword = process.env.SEED_ADMIN_PASSWORD;
  if (adminEmail && adminPassword) {
    const passwordHash = await bcrypt.hash(adminPassword, 10);
    await prisma.managerUser.upsert({
      where: { email: adminEmail.toLowerCase() },
      create: {
        email: adminEmail.toLowerCase(),
        passwordHash,
        name: "Admin",
        role: "ADMIN",
      },
      update: { passwordHash, role: "ADMIN" },
    });
    console.log(`Seeded admin user ${adminEmail}.`);
  }

  // Multi-manager form for production: SEED_MANAGERS is a JSON array of
  // { email, password, name, role? } - role defaults to "ADMIN". Passwords
  // never live in source; they're only ever passed in as an env var.
  const managersJson = process.env.SEED_MANAGERS;
  if (managersJson) {
    let managers: { email: string; password: string; name: string; role?: "ADMIN" | "MANAGER" }[];
    try {
      managers = JSON.parse(managersJson);
    } catch (err) {
      throw new Error(`SEED_MANAGERS is not valid JSON: ${err instanceof Error ? err.message : err}`);
    }
    for (const m of managers) {
      const passwordHash = await bcrypt.hash(m.password, 10);
      await prisma.managerUser.upsert({
        where: { email: m.email.toLowerCase() },
        create: {
          email: m.email.toLowerCase(),
          passwordHash,
          name: m.name,
          role: m.role ?? "ADMIN",
        },
        update: { passwordHash, name: m.name, role: m.role ?? "ADMIN" },
      });
      console.log(`Seeded manager user ${m.email}.`);
    }
  }

  if (!adminEmail && !managersJson) {
    console.log("No SEED_ADMIN_EMAIL or SEED_MANAGERS set - skipped creating any login.");
  }
}

main()
  .catch((err) => {
    console.error(err);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

File: backend/app/seed.py
"""Seeds the two initial users, the Egypt public holiday calendar, and a
starter set of job-role -> category rules. Safe to re-run (idempotent
upserts). Run with: python -m app.seed
"""

from datetime import date

from app.config import get_settings
from app.db import SessionLocal
from app.models import (
    EmployeeCategory,
    EmployeeNameOverride,
    Holiday,
    JobRoleCategoryRule,
    User,
    UserRole,
)
from app.security import hash_password

# Fixed-date holidays are confirmed by Egypt's official calendar. Islamic
# (Hijri) holidays depend on moon sighting and are estimates until confirmed
# closer to the date; they can shift by 1-2 days. Editable afterwards in
# Settings - this is just the starting seed.
EGYPT_HOLIDAYS_2026 = [
    ("2026-01-07", "Coptic Christmas"),
    ("2026-01-25", "January 25 Revolution / Police Day"),
    ("2026-03-20", "Eid al-Fitr (Day 1)"),
    ("2026-03-21", "Eid al-Fitr (Day 2)"),
    ("2026-03-22", "Eid al-Fitr (Day 3)"),
    ("2026-04-12", "Coptic Easter Sunday"),
    ("2026-04-13", "Sham El Nessim"),
    ("2026-04-25", "Sinai Liberation Day"),
    ("2026-05-01", "Labour Day"),
    ("2026-05-26", "Arafat Day"),
    ("2026-05-27", "Eid al-Adha (Day 1)"),
    ("2026-05-28", "Eid al-Adha (Day 2)"),
    ("2026-05-29", "Eid al-Adha (Day 3)"),
    ("2026-06-17", "Islamic New Year"),
    ("2026-06-30", "June 30 Revolution Day"),
    ("2026-07-23", "July 23 Revolution Day"),
    ("2026-08-26", "Prophet Muhammad's Birthday (Mawlid al-Nabi)"),
    ("2026-10-06", "Armed Forces Day"),
]

# Starter job-title -> category mapping. Add rows here (or via Settings)
# whenever a new job title shows up unassigned after a ZenHR sync.
STARTER_CATEGORY_RULES = {
    "General Manager": EmployeeCategory.MANAGEMENT,
    "Branch Manager": EmployeeCategory.MANAGEMENT,
    "Operations Manager": EmployeeCategory.MANAGEMENT,
    "Area Manager": EmployeeCategory.MANAGEMENT,
    "Supervisor": EmployeeCategory.MANAGEMENT,
    "Sales Representative": EmployeeCategory.SALES,
    "Sales Executive": EmployeeCategory.SALES,
    "Sales Support": EmployeeCategory.SALES_SUPPORT,
    "Collector": EmployeeCategory.COLLECTOR,
    "Delivery Agent": EmployeeCategory.DELIVERY_AGENT,
    "Co-Founder": EmployeeCategory.EXCLUDED,
    "HR Consultant": EmployeeCategory.EXCLUDED,
    "Managing Director": EmployeeCategory.EXCLUDED,
    # Exact ZenHR job-title strings observed on real new-hire syncs (rule
    # matching is an exact, case-sensitive dict lookup - see sync.py - so
    # these are kept separate from the title-cased entries above rather
    # than merged into them).
    "Delivery agent": EmployeeCategory.DELIVERY_AGENT,
    "Account Manager": EmployeeCategory.SALES,
    "Key Account Manager": EmployeeCategory.SALES,
    "Senior Account Manager": EmployeeCategory.SALES,
    "Logistics Supervisor": EmployeeCategory.MANAGEMENT,
    "HR Executive": EmployeeCategory.EXCLUDED,
    "office boy": EmployeeCategory.EXCLUDED,
    "office girl": EmployeeCategory.EXCLUDED,
}


# Reps whose Bricks account name differs from their ZenHR name, confirmed
# against the seven accounts the first live visits sync could not match.
# Stored trimmed: Bricks sends some of these with stray leading/trailing
# spaces, and matching normalizes whitespace on both sides anyway.
BRICKS_NAME_MAPPINGS: list[tuple[str, str, str]] = [
    ("115", "Kareem Elsobky", "Karim Ahmed"),
    ("119", "Mohamed Abdelrazek", "Mohamed Mohamed"),
    ("127", "Kareim Helmy", "Karim Mohamed"),
    ("135", "Amgad", "Abd Elrahman Elbannwy"),
    ("136", "Amr Mostafa", "Amr Hasan"),
    ("205", "Safaa Sabry", "Safaa Mohamed"),
    ("212", "Abdullah Ibrahim Habashi", "Abduallah Habashi"),
]


def seed_holidays(db) -> None:
    for iso_date, name in EGYPT_HOLIDAYS_2026:
        existing = db.query(Holiday).filter_by(holiday_date=date.fromisoformat(iso_date)).first()
        if existing:
            existing.name = name
        else:
            db.add(Holiday(holiday_date=date.fromisoformat(iso_date), name=name))


def seed_category_rules(db) -> None:
    for job_role, category in STARTER_CATEGORY_RULES.items():
        existing = db.get(JobRoleCategoryRule, job_role)
        if existing:
            existing.category = category
        else:
            db.add(JobRoleCategoryRule(job_role=job_role, category=category))


def seed_name_overrides(db) -> None:
    """Seed only what's missing: an admin who corrects a mapping in Settings
    keeps their version."""
    for employment_number, bricks_name, zenhr_name in BRICKS_NAME_MAPPINGS:
        if not db.query(EmployeeNameOverride).filter_by(employment_number=employment_number).first():
            db.add(
                EmployeeNameOverride(
                    employment_number=employment_number,
                    bricks_display_name=bricks_name,
                    note=f"ZenHR: {zenhr_name}",
                )
            )


def seed_users(db) -> None:
    settings = get_settings()
    if not db.query(User).filter_by(email=settings.seed_admin_email).first():
        db.add(
            User(
                email=settings.seed_admin_email,
                password_hash=hash_password(settings.seed_admin_password),
                name=settings.seed_admin_name,
                role=UserRole.ADMIN,
            )
        )


def main() -> None:
    db = SessionLocal()
    try:
        seed_users(db)
        seed_holidays(db)
        seed_category_rules(db)
        seed_name_overrides(db)
        db.commit()
        print("Seed complete.")
    finally:
        db.close()


if __name__ == "__main__":
    main()
